content/blog/vue-translate.md

```diff
@@ -0,0 +1,175 @@
+ commit({type: "CHANGE_LANGUAGE", payload: newLanguage})
+```
+
+### Listening for changes
+
+Listening for changes in a component is easy, it's baked into the framework. A simple translation component
+that just translates text and renders it in a span looks like this:
+
+```
+<template>
+    <span>{{translatedText}}</span>
+</template>
+
+<script>
+import {mapState} from Vuex
+
+export default {
+    props: ["translationKey"],
+    data() {
+        return {
+            translatedText: i18next.t(this.translationKey)
+        }
+    },
+    computed: mapState({ language -> state.language }),
+    watch: {
+        language() {
+            this.translatedText = i18next.t(this.translationKey)
+        }
+    }
+}
+
+</script>
+```
+
+The computed property "language" gets updated whenever the store state language changes, the watcher fires and the 
+text is updated.
+
+A directive has a completely different structure. There are 5 lifecycle events: `bind`, `inserted`, `update`, 
+`componentUpdated` and `unbind`. `bind` is fired when the component is first created, and the `update` hook allows us to listen for changes to the value of the directive. 
+It will fire if for example, the prop "title" changes in value in this component:
+
+```
+<template>
+    <h3 v-translate="title"></h3>
+</template>
+<script>
+export default {
+    props: ["title"]
+}
+</script>
+```
+
+But there is no directive hook that will fire when *store state* changes, i.e. our language. So we'll have to manage
+a subscription to the store ourselves. We can set up such a subscription on `bind`:
+
+```
+bind(el, binding) {
+     el.innerHTML = i18next.t(binding.value, {lng: store.state.language});
+     el.__lang_unwatch__ = store.watch(state => state.language, lng => {
+          el.innerHTML = i18next.t(binding.value, {lng});
+     })
+}
+```
+
+We have to be careful to remove the subscription when the element is destroyed, to prevent a proliferation 
+of watchers. We can do this on `unbind`:
+
+```
+unbind(el) {
+    el.__lang_unwatch__();
+}
+```
+
+To update the element when the directive value changes is straight-forward with the `update` hook, but we 
+also have to remove and recreate the store watcher, since the previous watcher has the initial value of the binding
+ cached within it:
+
+```
+update(el, binding) {
+    el.__lang_unwatch__();
+    bind(el, binding);
+}
+```
+
+Then we register our directive and it's good to go:
+
+```
+const translate = {
+    bind: bind,
+    update: update,
+    unbind: unbind
+}
+
+Vue.directive('translate', translate);
+```
+
+The above is the code for a directive that just translates text. The directive we created is able to 
+translate arbitrary attributes of an element as per:
+
+```<input v-translate:placeholder="'email'" />```
+
+You can see the full (typescript) code for the final directive
+ [here](https://github.com/mrc-ide/hint/blob/master/src/app/static/src/app/directives/translate.ts).
```

Comment:
I sent looking for the `changeLanguage` bit and could not find that in here (as earlier in the post) but found it in the store/root/actions.ts file - is that worth linking here too?

Reply:
It's a bit convoluted in our app now, so I've added a more complete code snippet above instead